sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -38,95 +260,208 @@ static kmac_error_t wait_status_bit(uint32_t bit_position, uint32_t bit_value) {
+  }
+  write_cnt += func_name_len;
 
-  err = wait_status_bit(KMAC_STATUS_SHA3_IDLE_BIT, 1);
+  // write left_encode of cust_str_len
+  err = write_left_encode(cust_str_len, prefix_buf_ptr + write_cnt, &write_cnt);
   if (err != kKmacOk) {
     return err;
   }
+  // copy cust_str
+  for (size_t i = 0; i < cust_str_len; ++i) {
+    prefix_buf_ptr[write_cnt + i] = cust_str[i];
+  }
+  write_cnt += cust_str_len;
+
+  // Fill the remaining bytes with 0x0
+  for (size_t i = write_cnt; i < 4 * ARRAYSIZE(prefix_buffer); ++i) {
+    prefix_buf_ptr[i] = 0x0;
+  }
+
+  // Copy from `prefix_buffer` to PREFIX_REGS
+  for (i = 0; i < KMAC_PREFIX_MULTIREG_COUNT; i++) {
+    abs_mmio_write32(kKmacBaseAddr + prefix_offsets[i], prefix_buffer[i]);
+  }
+
+  return kKmacOk;
+}
+
+OT_WARN_UNUSED_RESULT
+kmac_error_t kmac_init(kmac_operation_t operation,
+                       kmac_security_str_t security_str,
+                       const uint8_t *func_name, uint8_t func_name_len,
+                       const uint8_t *cust_str, size_t cust_str_len) {
+  if (cust_str_len && cust_str == NULL) {
+    return kKmacArgsError;
+  } else if (func_name_len && func_name == NULL) {
+    return kKmacArgsError;
+  }
+
+  if (cust_str_len + func_name_len > kKmacPrefixMaxSize) {
+    return kKmacArgsError;
+  }
+
+  kmac_error_t err = wait_status_bit(KMAC_STATUS_SHA3_IDLE_BIT, 1);
+  if (err != kKmacOk) {
+    return err;
+  }
+
+  uint32_t cfg_reg =
+      abs_mmio_read32(kKmacBaseAddr + KMAC_CFG_SHADOWED_REG_OFFSET);
 
-  cfg_reg = common_cfg_reg;
   switch (operation) {
-    case kSHAKE128:
-      cfg_reg =
-          bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD,
-                                 KMAC_CFG_SHADOWED_KSTRENGTH_VALUE_L128);
+    case kKmacOperationSHA3:
       cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD,
-                                       KMAC_CFG_SHADOWED_MODE_VALUE_SHAKE);
+                                       KMAC_CFG_SHADOWED_MODE_VALUE_SHA3);
       break;
-
-    case kSHAKE256:
-      cfg_reg =
-          bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD,
-                                 KMAC_CFG_SHADOWED_KSTRENGTH_VALUE_L256);
+    case kKmacOperationSHAKE:
       cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD,
                                        KMAC_CFG_SHADOWED_MODE_VALUE_SHAKE);
       break;
-
-    case kSHA3_224:
-      cfg_reg =
-          bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD,
-                                 KMAC_CFG_SHADOWED_KSTRENGTH_VALUE_L224);
+    case kKmacOperationKMAC:
+      cfg_reg = bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_KMAC_EN_BIT, 1);
       cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD,
-                                       KMAC_CFG_SHADOWED_MODE_VALUE_SHA3);
+                                       KMAC_CFG_SHADOWED_MODE_VALUE_CSHAKE);
```

Comment:
nit: Maybe add a comment here explaining that the KMAC IP encodes KMAC mode as CSHAKE mode plus KMAC_EN, since that's a little confusing when scanning this code.